schnorr_keypair.go

```diff
@@ -0,0 +1,165 @@
+func GeneratePrivateKey() (key *SchnorrKeyPair, err error) {
+	rawKey := SerializedPrivateKey{}
+	for {
+		n, err := rand.Read(rawKey[:])
+		if err != nil || n != len(rawKey) {
+			return nil, err
+		}
+		key, err = DeserializePrivateKey(&rawKey)
+		if err == nil {
+			return key, nil
+		}
+	}
+}
+
+// SerializePrivateKey returns the private key in the keypair.
+func (key *SchnorrKeyPair) SerializePrivateKey() *SerializedPrivateKey {
+	// TODO: Replace with upstream function when merged: https://github.com/bitcoin-core/secp256k1/pull/845
+	ret := SerializedPrivateKey{}
+	for i := 0; i < 32; i++ {
+		ret[i] = byte(key.keypair.data[i])
+	}
+	return &ret
+}
+
+// Add a tweak to the public key by doing `key + tweak % Group Order` and adjust the pub/priv keys according to parity. this adds it in place.
+// This is meant for creating BIP-32(HD) wallets
+func (key *SchnorrKeyPair) Add(tweak [32]byte) error {
+	if err := key.isZeroed(); err != nil {
+		return err
+	}
+	cPtrTweak := (*C.uchar)(&tweak[0])
+	ret := C.secp256k1_keypair_xonly_tweak_add(context, &key.keypair, cPtrTweak)
+	if ret != 1 {
+		return errors.New("failed Adding to private key. Tweak is bigger than the order or the complement of the private key")
+	}
+	return nil
+}
+
+// SchnorrPublicKey generates a PublicKey for the corresponding private key.
+func (key *SchnorrKeyPair) SchnorrPublicKey() (*SchnorrPublicKey, error) {
+	pubkey, _, err := key.schnorrPublicKeyInternal()
+	return pubkey, err
+}
+
+func (key *SchnorrKeyPair) schnorrPublicKeyInternal() (pubkey *SchnorrPublicKey, wasOdd bool, err error) {
+	if err := key.isZeroed(); err != nil {
+		return nil, false, err
+	}
+	pubkey = &SchnorrPublicKey{}
+	cParity := C.int(42)
+	ret := C.secp256k1_keypair_xonly_pub(context, &pubkey.pubkey, &cParity, &key.keypair)
+	if ret != 1 {
+		return nil, false, errors.New("the keypair contains invalid data")
+	}
+
+	return pubkey, parityBitToBool(cParity), nil
+
+}
+
+// SchnorrSign creates a schnorr signature using the private key and the input hashed message.
+// Notice: the [32] byte array *MUST* be a hash of a message.
+func (key *SchnorrKeyPair) SchnorrSign(hash *Hash) (*SchnorrSignature, error) {
+	var auxilaryRand [32]byte
+	n, err := rand.Read(auxilaryRand[:])
+	if err != nil || n != len(auxilaryRand) {
+		return nil, err
+	}
+	return key.schnorrSignInternal(hash, &auxilaryRand)
+}
+
+func (key *SchnorrKeyPair) schnorrSignInternal(hash *Hash, auxiliaryRand *[32]byte) (*SchnorrSignature, error) {
+	if err := key.isZeroed(); err != nil {
+		return nil, err
+	}
+	signature := SchnorrSignature{}
+	cPtrSig := (*C.uchar)(&signature.signature[0])
+	cPtrHash := (*C.uchar)(&hash[0])
+	cPtrAux := unsafe.Pointer(auxiliaryRand)
+	ret := C.secp256k1_schnorrsig_sign(context, cPtrSig, cPtrHash, &key.keypair, C.secp256k1_nonce_function_bip340, cPtrAux)
+	if ret != 1 {
+		return nil, errors.New("failed Signing. You should call `DeserializePrivateKey` before calling this")
```

Comment:
I don't understand this error

Reply:
That's a sign of a bad error phrasing :)

The only way for `schnorrsig_sign` to fail is if you somehow ended up with an invalid keypair, which can only happen if you wrote directly to `SchnorrKeyPair` instead of using the "constructor" `DeserializePrivateKey`